### PR TITLE
feat: add new filter support for search records endpoint

### PR DESCRIPTION
**Pull request by frascuchon**
# Description

This PR integrates the `POST /api/v1/dataset/:dataset_id/records/search`  with search engine to support `filter` and `sort` body attributes.

Also, the base filter classes are decorated as `dataclass`  in order to provide an easy way to compare args in tests (`dataclass` provide automatic `eq` method) 

Refs https://github.com/argilla-io/argilla/issues/4227

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [] New feature (non-breaking change which adds functionality)
- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**Checklist**

- [X] I added relevant documentation
- [X] I followed the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [X] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

### CODE REVIEW

File: src/argilla/server/schemas/v1/datasets.py [hunk 1]
```diff
@@ -787,7 +787,7 @@ class Order(BaseModel):
 
 
 class SearchRecordsQuery(BaseModel):
-    query: Optional[Query]
+    query: Optional[Query] = PydanticField(default_factory=Query)
```
Comment (jfcalvo) [inline src/argilla/server/schemas/v1/datasets.py]: Why is this change needed? Is `None` not a valid default value?

File: src/argilla/server/search_engine/base.py [hunk 2]
```diff
@@ -71,53 +71,54 @@
...
 class AndFilter:
-    def __init__(self, filters: List["Filter"]):
-        self.filters = filters
+    filters: List["Filter"]
 
 
 Filter = Union[AndFilter, TermsFilter, RangeFilter]
 
 
+@dataclasses.dataclass
```
Comment (jfcalvo) [inline src/argilla/server/search_engine/base.py:121]: You added this line but the class still has a `__init__` method. Did you miss to change this class?

**Review by jfcalvo — CHANGES_REQUESTED** [review]

File: src/argilla/server/schemas/v1/datasets.py (hunk 1 above)
Comment (frascuchon) [reply src/argilla/server/schemas/v1/datasets.py]: It's true that `None` is a valid value. I reverted the change and refactored a bit how this object is parsed inside the endpoints handlers.

File: src/argilla/server/search_engine/base.py (hunk 2 above)
Comment (frascuchon) [reply src/argilla/server/search_engine/base.py:121]: Yes. removed! Thanks

**Review by jfcalvo — APPROVED** [review]
